agent/mibgroup/hardware/fsys/fsys_mntent.c

```diff
@@ -176,7 +179,240 @@ _fsys_type( char *typename )
+                config_perror("Out of memory");
+                return;
+            }
+            strcpy(d_str, name);
+            *p = c;
+            di_curr->item_type = ITEM_STRING;
+            di_curr->item_details = (void *) d_str;
+            name = p;
+        }
+        if (!*name) {
+            di_curr->item_next = (conf_mount_item *) 0;
+            break;
+        }
+        di_curr->item_next =
+            (conf_mount_item *) malloc(sizeof(conf_mount_item));
+        if (!di_curr->item_next) {
+            SNMP_FREE(di_curr->item_next);
+            SNMP_FREE(d_new);
+            SNMP_FREE(di_curr);
+            SNMP_FREE(d_set);
+            SNMP_FREE(d_str);
+            config_perror("Out of memory");
+            return;
+        }
+        di_curr = di_curr->item_next;
+    }
+    d_new->list_next = conf_list;
+    conf_list = d_new;
+}
+
+static void
+free_mount_config(void)
+{
+    conf_mount_list *d_ptr = conf_list, *d_next;
+    conf_mount_item *di_ptr, *di_next;
+
+    while (d_ptr) {
+        d_next = d_ptr->list_next;
+        di_ptr = d_ptr->list_item;
+        while (di_ptr) {
+            di_next = di_ptr->item_next;
+            if (di_ptr->item_details)
+                free(di_ptr->item_details);
+            free((void *) di_ptr);
+            di_ptr = di_next;
+        }
+        free((void *) d_ptr);
+        d_ptr = d_next;
+    }
+    conf_list = (conf_mount_list *) 0;
+}
+
+static int
+match_mount_config_item(const char *name, conf_mount_item * di_ptr)
+{
+    int             result = 0;
+    size_t          len;
+    details_set    *d_set;
+    unsigned int    c;
+
+    if (di_ptr) {
+        switch (di_ptr->item_type) {
+        case ITEM_STRING:
+            len = strlen((const char *) di_ptr->item_details);
+            if (!strncmp(name, (const char *) di_ptr->item_details, len))
+                result = match_mount_config_item(name + len,
+                                                di_ptr->item_next);
+            break;
+        case ITEM_SET:
+            if (*name) {
+                d_set = (details_set *) di_ptr->item_details;
+                c = ((unsigned int) *name) & 0xff;
+                if ((*d_set)[c / 8] & (unsigned char) (1 << (c % 8)))
+                    result = match_mount_config_item(name + 1,
+                                                    di_ptr->item_next);
+            }
+            break;
+        case ITEM_STAR:
+            if (di_ptr->item_next) {
+                for (; !result && *name; name++)
+                    result = match_mount_config_item(name,
+                                                    di_ptr->item_next);
+            } else {
+                result = 1;
+            }
+            break;
+        case ITEM_ANY:
+            if (*name)
+                result = match_mount_config_item(name + 1,
+                                                di_ptr->item_next);
+            break;
+        }
+    } else {
```

Comment:
Please use libpcre instead of implementing regular expression matching from scratch. See also http://www.pcre.org/. See also the Net-SNMP --with_pcre configure option.